src/extractLevels.js

```diff
@@ -0,0 +1,141 @@
+import fs from 'fs'
+import IfcModel from './IfcModel.js'
+
+var ifcFilename = 'IFCFile'
+var relElementsArrAll = []
+var IFCSpacesELementsArrAll = [] 
+
+const USAGE = `Usage: node extractLevel.js <file.ifc>
+
+EXAMPLE
+
+node extractLevel.js index.ifc
+
+OUT:
+index.ifc_Level0.ifc
+index.ifc_Level1.ifc
+...
+
+`
+
+
+async function ExtractLevels(){
+    await ExtractRELID()
+    //console.log(relElementsArrAll)
+    await CreateManipulateNewIFC()
+}
+
+ExtractLevels()
+
+async function ExtractRELID(){
+
+    let args = process.argv.slice(2)
+    if (args.length < 1) {
+        console.error(USAGE)
+        return
+    }
+
+    ifcFilename = args[0]
+    console.log(ifcFilename)
+    const model = new IfcModel()
+    const rawFileData = fs.readFileSync(ifcFilename)
+    await model.open(rawFileData)
+    console.log(model.modelId);
+    const ifcRelSpatial = model.getEltsOfNamedType('IFCRELCONTAINEDINSPATIALSTRUCTURE')
+    const ifcSpace = model.getEltsOfNamedType('IFCSPACE')
+
+    
+    for (let i = 0; i< ifcRelSpatial.length; i++){
+        const relElements = ifcRelSpatial[i].RelatedElements;
+        const relElementsArr = []
+    
+        for (let j = 0; j < relElements.length; j++) {
+            relElementsArr[j] = relElements[j].value;
+        }
+
+        relElementsArrAll[i] = relElementsArr
+    }
+
+    for (let i = 0; i< ifcSpace.length; i++){
+        IFCSpacesELementsArrAll[i] =  ifcSpace[i].expressID
+    }
+    //console.log('FirstFunctionComplete');
+
+}
+
+//still writing this
+function GroupLevels(allLevels){
+    for (let i = 0; i<allLevels; i++){
+
+    }
+}
+
+async function CreateManipulateNewIFC(){
+
+    for (let i = 0; i<relElementsArrAll.length; i++){
+        var newfilename = ifcFilename+'_Level'+i+'.ifc'
+        await CopyFiles(ifcFilename, newfilename);
+        var expressIDsToRemoveIFCRel = relElementsArrAll[i];
+        var expressIDsToRemoveIFCSPACE = IFCSpacesELementsArrAll;
+        //console.log(expressIDsToRemove);
+        //console.log(newfilename);
+        await removeLinesFromIFCwithExpressID (newfilename, expressIDsToRemoveIFCRel)
+        await removeLinesFromIFCwithExpressID (newfilename, expressIDsToRemoveIFCSPACE)
+    }
+}
+
+
+function findLineWithExpressID (IFCdata, expressID){
+    var lines = IFCdata.split("\n");
```

Comment:
Can this split and search be moved to a single up-front operation, to setup an index of lines, and then do operations after that?

Reply:
Are we talking about maybe making a hashmap after one set of splits here? EXPRESSIDs being keys?

Reply:
Yep

```
const linesWithId = {}
...
const linesForId = linesWithId[id] || []
...
linesForId.push(line)
...
```

Reply:
Or actually, probably a set instead of array
```
const linesForId = linesWithId[id] || new Set()
...
linesForId.add(line)
```

Reply:
Addressed in updated version